src/lib/staged_ledger/staged_ledger.ml

```diff
@@ -1215,7 +1215,68 @@ module T = struct
       =
     (List.map ~f:(With_status.map ~f:Transaction.forget) a, b, c, d)
 
-  let check_commands ledger ~verifier (cs : User_command.t With_status.t list) =
+  type transaction_pool_proxy =
+    { find_by_hash :
+           Mina_transaction.Transaction_hash.t
+        -> Mina_transaction.Transaction_hash.User_command_with_valid_signature.t
+           option
+    }
+
+  let dummy_transaction_pool_proxy : transaction_pool_proxy =
+    { find_by_hash = const None }
+
+  (* NOTE: when we have access to txn pool, any txns in pool should be already
+     verified, hence we could utilize this fact to perform a faster version of
+     command verification *)
+  let verify_command_with_transaction_pool_proxy
+      (transaction_pool_proxy : transaction_pool_proxy)
+      (cmd_with_status : User_command.Verifiable.t With_status.t) =
+    let With_status.{ data = verifiable_cmd; _ } = cmd_with_status in
+    let cmd_hash =
+      verifiable_cmd |> User_command.of_verifiable
+      |> Transaction_hash.hash_command
```

Comment:
The only major comment for this PR: we need to check whether we could source transaction hash from outside of this function. Computing transaction hash is expensive and if possible we should avoid doing it and throwing it away, knowing that transaction hash might have been computed in other contexts for the same transaction.

Maybe in this case there are no such contexts at all, I don't know.